### PR TITLE
Add compartments to nutrition module

### PR DESCRIPTION
**Pull request by FrankApiyo**
This PR accomplishes the following:

1. Refactors `filterArgs` to use functions as opposed to objects that contain information about how to filter data
2. Generalize components such as `HierarchichalDataTable` and `DataCircleCard` to be able to work with each of the three modules
3. Fixes a few typos
4. Refactor `BasicInformation` to make it a more reusable component.
5. Replace the use of `ems` with `rems`
6. Fix the fonts issue by using fonts in the assets
7. Modify `VillageData` component shuch that a different set of tale columns are shown based on the module we currently are on.
8. Create a `ChildPatientDetail` component.

### CODE REVIEW

**Comment by cKellyDesign** [comment]
> This PR accomplishes the following:

@FrankApiyo do any of these points have tickets with more context?

File: src/containers/ChildPatientDetails/index.tsx [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+  private getRiskCartegorization = (): string => {
+    const reversedFilteredData: SmsData[] = [...this.state.filteredData];
+    reversedFilteredData.reverse();
+    for (const data in reversedFilteredData) {
+      if (reversedFilteredData[data].logface_risk) {
+        return reversedFilteredData[data].logface_risk;
+      }
+    }
+    return 'could not find any location';
```
Comment (cKellyDesign) [inline src/containers/ChildPatientDetails/index.tsx]: Let's move this to `constants`

File: src/containers/ChildPatientDetails/index.tsx [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+          </span>
+        </Link>
+        <div id="titleDiv">
+          <h2 id="patients_title">{PATIENT_DETAILS}</h2>
+        </div>
+        <Row>
+          <BasicInformation labelValuePairs={this.getBasicInformationProps()} />
+        </Row>
+        <Row id="filter-panel">
+          <p>Showing reports for:&emsp;</p>
```
Comment (cKellyDesign) [inline src/containers/ChildPatientDetails/index.tsx]: Move to `constants`? (no magic strings)

Comment (cKellyDesign) [reply src/containers/ChildPatientDetails/index.tsx]: And why include the `&emsp;`? Will there be a following string in the `<p>` at some point?

File: src/containers/ChildPatientDetails/index.tsx [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    );
+  }
+
+  private toggleDropdown() {
+    return false;
+  }
+
+  private getWeightsArray() {
+    return this.state.filteredData.map((smsData: SmsData) => {
+      return {
```
Comment (cKellyDesign) [inline src/containers/ChildPatientDetails/index.tsx]: Doc-strings for all these please 🙏

File: src/containers/Compartments/index.tsx [hunk 4]
```diff
@@ -35,162 +40,436 @@ import {
...
+
+  /**
+   * Here we determine the location id that the user is assiged to
+   * and use that to filter smsData in the store.
+   * @param props
+   * @param state
+   */
+  public static getDerivedStateFromProps(props: Props, state: State): State {
+    // add filter for this location here
+    const userLocationId = 'eccfe905-0e03-4188-98bc-22f141cccd0e';
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Is this required in this PR?

File: src/containers/Compartments/index.tsx [hunk 5]
```diff
@@ -35,162 +40,436 @@ import {
...
+    let filterFunction;
+    function locationDataIsAvailable() {
+      return props.villages.length && props.districts.length && props.communes.length;
+    }
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Could/should this be a util in helpers?

File: src/containers/Compartments/index.tsx [hunk 6]
```diff
@@ -35,162 +40,436 @@ import {
...
+    if (Compartments.isProvince(userLocationId, props.provinces) && locationDataIsAvailable()) {
+      filterFunction = (smsData: SmsData) => {
+        // tslint:disable-next-line: no-shadowed-variable
+        const village = props.villages.find(village => {
+          return village.location_id === smsData.location_id;
+        });
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Let's not break this rule and instead make the const name more contextually specific (like `provinceVillage` or something)

File: src/containers/Compartments/index.tsx [hunk 7]
```diff
@@ -35,162 +40,436 @@ import {
...
+      };
+    } else if (
+      Compartments.isDistrict(userLocationId, props.districts) &&
+      locationDataIsAvailable()
+    ) {
+      filterFunction = (smsData: SmsData) => {
+        // tslint:disable-next-line: no-shadowed-variable
+        const village = props.villages.find(village => {
+          return village.location_id === smsData.location_id;
+        });
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Same as above and like `districtVillage`

File: src/containers/Compartments/index.tsx [hunk 8]
```diff
@@ -35,162 +40,436 @@ import {
...
+    return districts.find((district: Location) => district.location_id === locationId);
+  };
+
+  private static isCommune = (locationId: string, communes: Location[]) => {
+    return communes.find((commune: Location) => commune.location_id === locationId);
+  };
+
+  private static isVillage = (locationId: string, villages: Location[]) => {
+    return villages.find((village: Location) => village.location_id === locationId);
+  };
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Doc-strings?

File: src/containers/Compartments/index.tsx [hunk 9]
```diff
@@ -35,162 +40,436 @@ import {
...
-              {this.props.module === NBC_AND_PNC ? (
-                <ConnectedDataCircleCard
-                  {...dataCircleCardTestProps}
-                  module={this.props.module}
-                  className={'invisible-but-visible'}
-                />
-              ) : null}
-            </CardGroup>
-          </div>
+          <React.Fragment>
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Noice 👍

File: src/containers/Compartments/index.tsx [hunk 10]
```diff
@@ -233,16 +520,25 @@ class Compartments extends Component<Props, {}> {
...
+    communes: getLocationsOfLevel(state, 'Commune'),
     dataFetched: smsDataFetched(state),
+    districts: getLocationsOfLevel(state, 'District'),
+    filterArgsInStore: getFilterArgs(state),
+    provinces: getLocationsOfLevel(state, 'Province'),
     smsData: getFilterArgs(state)
-      ? getFilteredSmsData(state, getFilterArgs(state) as FilterArgs[])
+      ? getFilteredSmsData(state, getFilterArgs(state) as Array<(smsData: SmsData) => boolean>)
       : getSmsData(state),
+    villages: getLocationsOfLevel(state, 'Village'),
```
Comment (cKellyDesign) [inline src/containers/Compartments/index.tsx]: Let's move these magic strings to `constants` pls

File: src/containers/HierarchichalDataTable/index.tsx [hunk 11]
```diff
@@ -318,37 +322,30 @@ class HierarchichalDataTable extends Component<Props, State> {
...
   public render() {
+    const villageData = this.props.smsData.filter((dataItem: SmsData) => {
+      const locationIds = this.state.data.map((location: LocationWithData) => location.location_id);
+      return (
+        locationIds.includes(dataItem.location_id) &&
+        (this.props.risk_highligter
+          ? dataItem.logface_risk.includes(this.props.risk_highligter)
+          : true)
+      );
+    });
```
Comment (cKellyDesign) [inline src/containers/HierarchichalDataTable/index.tsx]: Does this potentially get expensive to compute? If so, should it be moved to `getDerivedStateFromProps` or `mapStateToProps`?

File: src/containers/pages/ModuleHome/index.css [hunk 12]
```diff
@@ -13,12 +13,10 @@
   max-width: 25%;
   min-width: 140px;
   flex: 1;
-  font-family: 'Trebuchet MS', 'Helvetica';
```
Comment (cKellyDesign) [inline src/containers/pages/ModuleHome/index.css]: Yikes, at least `Roboto` and `Helvetica` are decent fonts, ha

File: src/helpers/utils.tsx [hunk 13]
```diff
@@ -57,3 +59,65 @@ export const sortFunction = (firstE1: SmsData, secondE1: SmsData): number => {
...
+  return districtId
+    ? districts.find((location: Location) => location.location_id === districtId)!.parent_id
+    : null;
+};
+
+export const filterByPatientAndSort = (props: {
+  patientId: string;
+  smsData: SmsData[];
+}): SmsData[] => {
+  return props.smsData
```
Comment (cKellyDesign) [inline src/helpers/utils.tsx]: Doc-strings

File: src/helpers/utils.tsx [hunk 14]
```diff
@@ -57,3 +59,65 @@ export const sortFunction = (firstE1: SmsData, secondE1: SmsData): number => {
...
+      if (event1.EventDate > event2.EventDate) {
+        return 1;
+      }
+      return 0;
+    });
+};
+
+export const getNumberOfDaysSinceDate = (date: string): number => {
+  return Math.floor((new Date().getTime() - new Date(date).getTime()) / (1000 * 3600 * 24));
+};
```
Comment (cKellyDesign) [inline src/helpers/utils.tsx]: Doc-strings

File: src/store/ducks/sms_events.ts [hunk 15]
```diff
@@ -169,61 +166,34 @@ export function smsDataFetched(state: Partial<Store>): boolean {
...
  * of a field specified.
  * @param {Partil<Store>} state - the state of the SmsReducer redux store
  * @param {field} string - the name of the field to filter by
  * @param {value} string | number - the string or number value of the field specified
  */
-export function getFilteredSmsData(state: Partial<Store>, filterArgs: FilterArgs[]): SmsData[] {
+export function getFilteredSmsData(
+  state: Partial<Store>,
+  filterArgs: Array<(smsData: SmsData) => boolean>
+): SmsData[] {
```
Comment (cKellyDesign) [inline src/store/ducks/sms_events.ts]: Updated doc-strings?

File: src/store/ducks/sms_events.ts [hunk 16]
```diff
@@ -169,61 +166,34 @@ export function smsDataFetched(state: Partial<Store>): boolean {
...
-      return actualValue < targetValue;
-    default:
-      return false;
-  }
-}
-
 /** Returns the filterArgs currently in the store */
-export function getFilterArgs(state: Partial<Store>): FilterArgs[] | null {
-  return (state as any)[reducerName].filterArgs;
+export function getFilterArgs(state: Partial<Store>): Array<(smsData: SmsData) => boolean> {
```
Comment (cKellyDesign) [inline src/store/ducks/sms_events.ts]: Doc-strings including `@param`s and `@return`

File: src/App/Routes.tsx [hunk 17]
```diff
@@ -125,13 +136,13 @@ export const Routes = (props: RoutesProps) => {
...
-                    field: SMS_TYPE,
-                    value: PREGNANCY_REGISTRATION,
-                  },
-                ]}
+                filterArgs={
+                  [
+                    (smsData: SmsData) => {
+                      return smsData.sms_type === PREGNANCY_REGISTRATION;
+                    },
+                  ] as Array<(smsData: SmsData) => boolean>
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/App/Routes.tsx]: yeah! probably should.

**Comment by FrankApiyo** [comment]
> > This PR accomplishes the following:
> 
> @FrankApiyo do any of these points have tickets with more context?

@cKellyDesign sorry, :cry: they don't

File: src/App/Routes.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/App/Routes.tsx]: done

File: src/components/DataCircleCard/index.tsx [hunk 18]
```diff
@@ -35,14 +57,22 @@ function DataCircleCard({
...
     <Card className={`dataCircleCard ${className}`}>
       <CardTitle>{title}</CardTitle>
       <CardBody>
         <ul className="circlesRow">
           <li className="red">
             <Link
-              to={`${HIERARCHICAL_DATA_URL}/${module}/${HIGH}/${title}`}
+              to={`${getModuleLink(
+                module
+              )}${HIERARCHICAL_DATA_URL}/${module}/${HIGH}/${title}/${level}/down/${locationId}/${permissionLevel}`}
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/DataCircleCard/index.tsx]: I does need do be included in this PR but there's no ticket for it.

File: src/components/DataCircleCard/index.tsx [hunk 19]
```diff
@@ -20,11 +27,26 @@ interface Props {
...
   title: string;
   addFilterArgsActionCreator?: typeof addFilterArgs;
-  filterArgs?: FilterArgs[];
-  module: string;
+  filterArgs?: Array<(smsData: SmsData) => boolean>;
+  module: PREGNANCY | NBC_AND_PNC_CHILD | NBC_AND_PNC_WOMAN | NUTRITION | '';
   className?: string;
 }
 
+export function getModuleLink(module: string) {
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/DataCircleCard/index.tsx]: done.

**Comment by FrankApiyo** [comment]
About updating doc-strings for the functions in the locations reducer module and the functions in utils... this has been done,  comprehensively,  in a separate PR/branch: https://github.com/OpenSRP/opensrp-web/pull/106

File: src/components/BasicInformation/index.tsx [hunk 20]
```diff
@@ -29,36 +21,36 @@ export default function BasicInformation(props: Props) {
...
+                    return (
+                      <tr key={index}>
+                        <td>{labelValuePair.label}</td>
+                        <td>{labelValuePair.value}</td>
+                      </tr>
+                    );
+                  } else {
+                    return null;
+                  }
+                })}
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/BasicInformation/index.tsx]: they really aren't... we just want some to be on the left and some on the right....

File: src/components/DataCircleCard/index.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/DataCircleCard/index.tsx]: Ok cool, if it needs to be in this PR then we can skip the ticket as long as we get it implemented

File: src/components/VillageData/index.tsx [hunk 21]
```diff
@@ -114,4 +120,62 @@ export default class VillageData extends React.Component<Props, State> {
...
+        <td className="default-width">{getNumberOfDaysSinceDate(dataItem.EventDate)}</td>
+        <td className="default-width">{dataItem.health_worker_location_name}</td>
+        <td className="default-width">{dataItem.mother_symptoms}</td>
+        <td className="default-width">{dataItem.health_worker_location_name}</td>
+        <td className="default-width">
+          <RiskColoring {...{ risk: dataItem.logface_risk }} />
+        </td>
+      </tr>
+    );
+  };
```
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/VillageData/index.tsx]: @FrankApiyo does this have tests?

File: src/components/WeightAndHeightChart/index.tsx [hunk 22]
```diff
@@ -72,12 +94,17 @@ export default class MotherWeightChart extends React.Component<Props, State> {
...
           },
           formatter() {
-            return `Mother's weight - ${this.x}<br> weight <b>${this.y}</b>`;
+            return `${WeightAndHeightChart.legendString} - ${(this.x + '').slice(
+              0,
+              (this.x + '').lastIndexOf(' ')
+            )}<br>${WeightAndHeightChart.xAxisLabel}  <b>${this.y}</b> ${
+              WeightAndHeightChart.units
+            }`;
           },
```
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/WeightAndHeightChart/index.tsx]: @FrankApiyo ☝️

File: src/components/WeightAndHeightChart/test/index.test.tsx [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+  it('must render correctly', () => {
+    const wrapper = mount(
+      <Provider store={store}>
+        <Router history={history}>
+          <WeightAndHeightChart weights={weights} {...defaultProps} />
+        </Router>
+      </Provider>
+    );
+    expect(toJson(wrapper.find('#wrapper-id'))).toMatchSnapshot();
+  });
```
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/WeightAndHeightChart/test/index.test.tsx]: @FrankApiyo does this have tests?

**Comment by cKellyDesign** [comment]
> this has been done, comprehensively, in a separate PR/branch: #103

@FrankApiyo cool I will cross reference with #103 and resolve the threads that are covered there

File: src/containers/Compartments/index.tsx (hunk 8 above)
Comment (FrankApiyo) [reply src/containers/Compartments/index.tsx]: These have been refactored here:  #103 and moved to the `utils` and doc-strings added.

File: src/components/VillageData/index.css [hunk 24]
```diff
@@ -26,3 +26,153 @@
...
+.compartment-data-table .high {
+  background-color: #ffe8e6;
+}
+
+.compartment-data-table .low {
+  background-color: #ffe9db;
+}
+
+.compartment-data-table .no {
+  background-color: rgba(55, 174, 18, 0.10000000149011612);
```
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/VillageData/index.css]: I'm going to have to git blame this one.

[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/VillageData/index.css]: I saw it a few places so at least it's consistent (at least I think it's consistent)

[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/VillageData/index.css]: It was me :smile: . I got this value from `marvel`

File: src/helpers/utils.tsx (hunk 14 above)
Comment (cKellyDesign) [reply src/helpers/utils.tsx]: https://github.com/OpenSRP/opensrp-web/pull/103/files#diff-599ee30dbf9c78dae505330fc1ae3bd9R269-R275

File: src/helpers/utils.tsx (hunk 13 above)
Comment (cKellyDesign) [reply src/helpers/utils.tsx]: https://github.com/OpenSRP/opensrp-web/pull/103/files#diff-599ee30dbf9c78dae505330fc1ae3bd9R246-R252

File: src/containers/ChildPatientDetails/index.tsx (hunk 3 above)
Comment (cKellyDesign) [reply src/containers/ChildPatientDetails/index.tsx]: I don't see this in 103 so assume it still needs ☝️

File: src/components/VillageData/index.css (hunk 24 above)
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/VillageData/index.css]: Makes sense and suppose it's not a big deal really :P

File: src/components/VillageData/index.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/VillageData/index.tsx]: Nope.. working on it.

[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/VillageData/index.tsx]: done and dusted

File: src/containers/Compartments/index.tsx (hunk 4 above)
Comment (FrankApiyo) [reply src/containers/Compartments/index.tsx]: #103 removes this...

File: src/containers/Compartments/index.tsx (hunk 5 above)
Comment (FrankApiyo) [reply src/containers/Compartments/index.tsx]: sure thing.

Comment (FrankApiyo) [reply src/containers/Compartments/index.tsx]: done!

File: src/containers/Compartments/index.tsx (hunk 10 above)
Comment (FrankApiyo) [reply src/containers/Compartments/index.tsx]: done!

File: src/containers/ChildPatientDetails/index.tsx (hunk 3 above)
Comment (FrankApiyo) [reply src/containers/ChildPatientDetails/index.tsx]: @p-netm I'm pretty sure this is a rear moment when @cKellyDesign is confusing you for me. I'm pretty sure. :rofl:

File: src/containers/ChildPatientDetails/index.tsx (hunk 1 above)
Comment (FrankApiyo) [reply src/containers/ChildPatientDetails/index.tsx]: done!

File: src/containers/ChildPatientDetails/index.tsx (hunk 2 above)
Comment (FrankApiyo) [reply src/containers/ChildPatientDetails/index.tsx]: we need some breathing room between the `<p>` tag and and the `<Dropdown>` that follows there. I'll remove that and add some padding instead.

File: src/containers/HierarchichalDataTable/index.tsx (hunk 11 above)
Comment (FrankApiyo) [reply src/containers/HierarchichalDataTable/index.tsx]: done!

File: src/containers/ChildPatientDetails/index.tsx (hunk 3 above)
Comment (FrankApiyo) [reply src/containers/ChildPatientDetails/index.tsx]: done!

File: src/components/WeightAndHeightChart/test/index.test.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/WeightAndHeightChart/test/index.test.tsx]: I thinks we should.... on it.

File: src/components/WeightAndHeightChart/index.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/WeightAndHeightChart/index.tsx]: yes

File: src/components/WeightAndHeightChart/test/index.test.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/WeightAndHeightChart/test/index.test.tsx]: Might not be able to do this in this PR. I'll create a separate issue for this.

[the remark this answers is not in the document]
Comment (FrankApiyo) [reply src/components/WeightAndHeightChart/test/index.test.tsx]: issue created here: https://github.com/OpenSRP/opensrp-web/issues/111

File: src/components/WeightAndHeightChart/index.tsx (hunk 22 above)
[the remark this answers is not in the document]
Comment (cKellyDesign) [reply src/components/WeightAndHeightChart/index.tsx]: Why not `String(this.x)`?

**Pull request merged**